_posts/2021-09-23-unused-beans.adoc

```diff
@@ -0,0 +1,73 @@
+---
+layout: post
+title: 'Unused Beans and Why We Remove Them'
+date: 2021-09-23
+tags: cdi arc 
+synopsis: Why does Quarkus detect and remove unused beans by default?
+author: mkouba
+---
+
+Quarkus is a build-time oriented stack, i.e. it tries to do as much as possible at build time to improve the startup time and memory usage of your application.
+However, it's not always so easy and straightforward.
+Many existing frameworks and libraries do not take this design choice into account.
+And CDI is one of them.
+
+== The Why
+
+A traditional CDI container attempts to find all beans during the application bootstrap.
+For each bean a metadata object is created and kept for the lifetime of the application.
+This allows for certain degree of dynamicity at runtime but it's suboptimal when it comes to memory consumption and application startup time.
+Quarkus, on the other hand, attempts to detect and remove all _unused_ beans, interceptors and decorators during build by default.
+And the reason is simple.
+This optimization helps to minimize the amount of generated classes and the number of metadata objects used at runtime, thus conserving memory.
+
+Speaking of generated classes.
+The bean discovery, validation and wiring all components together - all this is perfomed at build time.
+Quarkus then stores the collected metadata in the bytecode, i.e. for each bean one or more classes are generated.
+In order to fullfill some basic CDI API requirements, a bean has at least a corresponding `javax.enterprise.inject.spi.Bean` implementation.
+If it's a normal scoped bean then a client proxy class must be also generated.
+Finally, intercepted and decorated beans require some more internal constructs.
+
+Imagine that your application contains 50 beans that are actually not used anywhere.
+If they're normal scoped (e.g. has `@ApplicationScoped`) and intercepted (e.g. annotated with `@Transactional`) you can expect more than 150 generated classes.
+And these classes are completely useless.
+Still, the container would have to instantiate and hold a reference on those 50+ bean metadata classes.
+Needless to say, that the bean classes and any referenced classes cannot be a subject of dead code elimination when a native image is generated.
+Therefore, Quarkus implements an algorithm to get rid of all these classes.
+
+== What’s Actually Removed?
+
+Quarkus first identifies so-called _unremovable_ beans that form the roots in the dependency tree.
+These classes very often represent the application _entry points_.
+A good example is a JAX-RS resource class.
+Another example is a bean which declares a `@Scheduled` method.
+
+An _unremovable_ bean:
+- is excluded from removal by an extension, or
```

Comment:
I assume that "an extension" here means Quarkus extension, and so this case covers the examples from previous paragraph. If so, I think this connection is far from obvious. I'd rephrase to something like this:

> Quarkus first identifies so-called _unremovable_ beans that form the roots in the dependency tree.
> Unremovable beans are beans that:
> - declare an observer method, or
> - have a name designated via `@Named`, or
> - are otherwise identified by Quarkus extensions.
>
> The last point is probably most important, because this is how application _entry points_ are made unremovable.
> A good example is a JAX-RS resource class, identified by the RESTEasy extension.
> Another example is a bean which declares a `@Scheduled` method, identified by the Scheduler extension.